pypose/module/icp.py

```diff
@@ -0,0 +1,102 @@
+
+    4. The algorithm continues to iterate through these steps until the change in the
+    calculated distance falls below the specified tolerance level or the maximum number
+    of iteration steps is reached.
+
+    Example:
+        >>> import torch, pypose as pp
+        >>> source = torch.tensor([[[0., 0., 0.],
+        ...                         [1., 0., 0.],
+        ...                         [2.,  0, 0.]]])
+        >>> target = torch.tensor([[[0.2,      0.1,  0.],
+        ...                         [1.1397, 0.442,  0.],
+        ...                         [2.0794, 0.7840, 0.]]])
+        >>> icp = pp.module.ICP()
+        >>> icp(source, target)
+        SE3Type LieTensor:
+        LieTensor([[0.2000, 0.1000, 0.0000, 0.0000, 0.0000, 0.1736, 0.9848]])
+
+    Warning:
+        It's important to note that the solution is sensitive to the initialization.
+    '''
+    def __init__(self, steps=200, tol=1e-6, init=None):
+        super().__init__()
+        assert init is None or is_SE3(init), "The initial transformation is not SE3Type."
+        self.steps, self.tol, self.init = steps, tol, init
+
+    def forward(self, source, target, ord=2, dim=-1, init=None):
+        r'''
+        Args:
+            source (``torch.Tensor``): the source point clouds with shape
+                (..., points_num, 3).
+            target (``torch.Tensor``): the target point clouds with shape
+                (..., points_num, 3).
+            ord (``int``, optional): the order of norm to use for distance calculation.
+                Default: ``2`` (Euclidean distance).
+            dim (``int``, optional): the dimension encompassing the point cloud
+                coordinates, utilized for calculating distance.
+                Default: ``-1`` (The last dimension).
+            init (``LieTensor``, optional): the initial transformation :math:`T_{init}` in
+                ``SE3type``. If not ``None``, it will suppress the ``init`` given by the
+                class constructor. Default: ``None``.
+
+        Returns:
+            ``LieTensor``: The estimated transformation (``SE3type``) from source to
+            target point cloud.
+        '''
+        temporal, errlast = source, 0
+        init = init if init is not None else self.init
+        if init is not None:
+            assert is_SE3(init), "The initial transformation is not SE3Type."
+            temporal = init.unsqueeze(-2) @ temporal
+        for _ in range(self.steps):
+            knndist, knnidx = knn(temporal, target, k=1, ord=ord, dim=dim)
+            errnew = knndist.squeeze(-1).mean(dim=-1)
+            if torch.all((errnew - errlast).abs() < self.tol):
+                break
+            errlast = errnew
+            target = target.view(*([1] * (knnidx.dim() - target.dim())), *target.shape)
+            target = target.expand(*knnidx.shape[:-2], target.size(-2),target.size(-1))
+            knntarget = torch.gather(target, -2, knnidx.expand(*knnidx.shape[:-1],
```

Comment:
How about out of the loop, we define
```python
batch = torch.broadcast_shapes(source.shape[:-2], target.shape[:-2])
```
then replace these two line by
```python
target = target.expand(batch + target.shape[-2:])
knntarget = torch.gather(target, -2, knnidx.expand(batch + source.shape[-2:]))
```

Reply:
It looks much more elegant now!